### ISSUE TITLE
Some question about Automatic Differentiation

### ISSUE THREAD

**Issue by xichaoqiang** [fpicetti/occamypy#29]
Dear Dr.Picetti @fpicetti 

Thanks for sharing tutorials about automatic differentiation.

These tutorials are very interesting.

I have some question.

1.For my understating, I only need define the forwarding operator and cost function with automatic differentiation. Is it right?

2. How can we use automatic differentiation with nonlinear conjugate gradient method?

3.Would you like add some more tutorials about VSP(1D), tomography (2D) inversion with automatic differentiation? I think the implements are very helpful.

Best regards,
Xi

**Comment by fpicetti** [comment]
Hello @xichaoqiang,

Thanks for reaching out.

Automatic Differentiation is a way to compute complex gradients by storing and back-propagating the single gradients in a computation graph.

1. In the [LS-RTM example with AD](https://github.com/fpicetti/occamypy/blob/master/tutorials/2D%20LS-RTM%20with%20devito%20and%20Automatic%20Differentiation.ipynb), I show how to cast a occamypy.Operator to a function that can be utilized by pytorch's autograd engine. You need to define **both the forward and the adjoint methods**, as they will be called in the forward "apply" and in the backprop, respectively. In other words, this tutorial is useful if you need to use occamypy operators in a machine learning framework.
In the [second AD tutorial](https://github.com/fpicetti/occamypy/blob/master/tutorials/Automatic%20Differentiation%20for%20nonlinear%20operators.ipynb), which is just a draft, I show how to embed a torch-native nonlinear forward function to a "operator-like" class, which implements the adjoint as the Jacobian-vector product provided by the `functorch` package. It is quite limited to squared operators; there's lot to study in order to generalize it to any kind of operator.
2. How to use AD with NLCG solver? In my vision, you need to use the `OperatorAD` class implemented in the second tutorial, wrap it to a nonlinear operator and set up your `problem`. However, I believe there's a smarter way to approach such optimization. My first pick would be to use just pytorch and its SGD. The point of AD is not to implement complex jacobians, and pytorch allows us to skip this.
3. All the tutorials were born by our ongoing researches; I don't foresee in the near future any other tutorials on VSP and TT with automatic differentiation. I do believe that it should be easy to take inspiration from our tutorials, and I encourage you to try. Let us know how it goes!

Bests.

**Comment by xichaoqiang** [comment]
Thanks for your kindly reply.
